File: src/main/java/org/usfirst/irs1318/gamesim/engine/objective/TaskDetails.java
package org.usfirst.irs1318.gamesim.engine.objective;

import java.time.Duration;

public final class TaskDetails {
    private final Duration taskDuration;
    private final Duration reactionDuration;

    /**
     * A number between 0 and 1.
     */
    private final float reliability;

    private final Task task;

    public TaskDetails(Duration taskDuration, Duration reactionDuration, float reliability, Task task) {
        this.taskDuration = taskDuration;
        this.reactionDuration = reactionDuration;
        this.reliability = reliability;
        this.task = task;
    }

    /**
     * @return true if a rng is <= reliability.
     */
    public boolean succeeded() {
        // TODO
        return true;
    }

    public Duration getTaskDuration() {
        return taskDuration;
    }

    public Duration getReactionDuration() {
        return reactionDuration;
    }

    public float getReliability() {
        return reliability;
    }

    public Task getTask() {
        return task;
    }
}

File: src/main/scala/org/usfirst/irs1318/gamesim/engine/objective/TaskDetails.scala
package org.usfirst.irs1318.gamesim.engine.objective

case class TaskDetails(taskDuration: Long,
                       reactionDuration: Long,
                       reliability: Float,
                       task: Task)

object TaskDetails {
  case class Builder(taskDetails: TaskDetails) {
    def this() = this(TaskDetails(0, 0, 1.0f, new Task.Builder().build))
    @inline def mapTaskDetails(fn: TaskDetails => TaskDetails): Builder = copy(taskDetails = fn(taskDetails))
    def setTaskDuration(taskDuration: Long): Builder = mapTaskDetails(_.copy(taskDuration = taskDuration))
    def setReactionDuration(reactionDuration: Long): Builder = mapTaskDetails(_.copy(reactionDuration = reactionDuration))
    def setReliability(reliability: Float): Builder = mapTaskDetails(_.copy(reliability = reliability))
    def setTask(task: Task): Builder = mapTaskDetails(_.copy(task = task))
    def build: TaskDetails = taskDetails
  }
}
